sperkovic

soperkov opened the pull request:
skoro pa riješena vježba

soperkov committed:
sperkovic

rijesena vjezba i dodano resize-anje slika (nije bas najbolje jer jako veliku sliku ne resize-a dobro pa mozda se pozabavim time, i mozda napravim matodu za provjeru resize-anja slike)

soperkov committed:
nisu pokriveni svi slučajevi u resize-anju ali potrudila sam se barem većina da bude. molim vas pomoć oko erase_bkgnd jer nisam uspjela izgubiti blinkanje pa sam samo obrisala implementaciju.

ncaklovic commented on main.cpp:
```diff
@@ -2,16 +2,89 @@
+	RectF layout_rect(static_cast<float>(rc.left), 0.0f,
+		static_cast<float>(rc.right - rc.left),
+		static_cast<float>(rc.bottom - rc.top));
+	layout_rect.Y = static_cast<float>(rc.bottom) - my_font.GetHeight(&g);
```
Ne treba vam uopće _font height_ ako koristite i [vertical alignment](https://learn.microsoft.com/en-us/windows/win32/api/gdiplusstringformat/nf-gdiplusstringformat-stringformat-setlinealignment)

ncaklovic commented on main.h:
```diff
@@ -18,8 +18,15 @@ class gdiplus_application : public vsite::nwp::application
 class main_window : public vsite::nwp::window {
+	Gdiplus::Image* img = nullptr;
+	TCHAR* text;
+	HDC hdc;
```
`hdc` valjda ne treba kao member

ncaklovic commented on main.h:
```diff
@@ -18,8 +18,15 @@ class gdiplus_application : public vsite::nwp::application
 
 class main_window : public vsite::nwp::window {
+	Gdiplus::Image* img = nullptr;
+	TCHAR* text;
```
valjda tstring (ili barem std::wstring).

ncaklovic commented on main.cpp:
```diff
@@ -2,16 +2,89 @@
+			if (::GetOpenFileName(&ofn)) {
+				img = check_scaling(*this, Image::FromFile(path));
+
+				std::wstring filePath = std::filesystem::path(path).filename().wstring();
```
zar nije dovoljno `text = std::filesystem::path(path).filename()`

soperkov replied:
koliko se sjećam nije radilo na vježbi, crvenilo se.

ncaklovic commented on main.cpp:
```diff
@@ -2,16 +2,89 @@
+	GetClientRect(*this, &rc);
+	Graphics g(hdc);
+	g.DrawImage(img, 0, 0, rc.right, rc.bottom);
+	draw_string(hdc, text, rc);
```
promijenite da `draw_string` prima `const TCHAR*` pa ćete moći poslati `text.c_str()`

ncaklovic commented on main.cpp:
```diff
@@ -2,16 +2,89 @@
+
+void draw_string(HDC hdc, TCHAR* text, RECT rc) {
+
+	Graphics g(hdc);
```
Hm, ovo se zove iz `on_paint` gdje već živi jedan objekt tipa `Graphics`, nisam siguran može li tu biti problema.

ncaklovic commented on main.cpp:
```diff
@@ -2,16 +2,89 @@
+			ofn.lpstrFilter = filter;
+			ofn.Flags = OFN_HIDEREADONLY;
+			if (::GetOpenFileName(&ofn)) {
+				img = check_scaling(*this, Image::FromFile(path));
```
Pazite, ako je slika već otvorena ovdje ne oslobađate prethodnu.
Treba vam `delete` ili (još bolje) `unique_ptr` kao member.

ncaklovic commented on main.cpp:
```diff
@@ -2,16 +2,89 @@
+
+	img_width = static_cast<int>(img_width * scale_factor);
+	img_height = static_cast<int>(img_height * scale_factor);
+	return img->GetThumbnailImage(img_width, img_height);
```
Ne vidim potrebu za ovim, pa možete samo izračunati veličinu prozora (ako već želite), ionako se slika crta po cijelom prozoru (koliki god bio, a može mu se mijenjati veličina).

soperkov replied:
htjela sam da mi se prozor resize-a po slici, a ako je slika veća od desktopa da se ona smanji tako da stane na ekran, da ga ne premašuje, zato sam ovo sve računala

ncaklovic commented on main.h:
```diff
@@ -18,8 +18,15 @@ class gdiplus_application : public vsite::nwp::application
 	void on_paint(HDC hdc) override;
 	void on_command(int id) override;
 	void on_destroy() override;
+	bool on_erase_bkgnd(HDC hdc) override;
```
Kad se već slika crta preko cijelog prozora, možete izbjeći brisanje pozadine background brush-om.
Samo prepišite funkciju i vratite `true`, pogledajte [help za WM_ERASEBKGND](https://learn.microsoft.com/en-us/windows/win32/winmsg/wm-erasebkgnd) i [handling u baznoj klasi](https://github.com/vsite-nwp/v07/blob/main/nwpwin.cpp#L91).

soperkov committed:
evo nesto sa bitmapom umjesto image*, čini mi se lošija slika (više nekih nepravilnosti) ali barem ekran ne bljeska dok prikazuje sliku, mislim da je riješeno ovo iz komentara

ncaklovic commented:
Dodajte `SetStretchBltMode(hdc, COLORONCOLOR);` prije `StretchBlt()`

soperkov committed:
dodan SetStretchBltMode

ncaklovic commented:
Ne treba vam `DrawImage` da biste dobili `HBITMAP`. Klasa `Bitmap` ima [konstruktor sa pathom](https://learn.microsoft.com/en-us/windows/win32/api/gdiplusheaders/nf-gdiplusheaders-bitmap-bitmap(constwchar_bool)) (ima i [FromFile](https://learn.microsoft.com/en-us/windows/win32/api/gdiplusheaders/nf-gdiplusheaders-bitmap-fromfile)).
Ne trebaju vam `GetObject` pozivi, `Bitmap` je izvedena iz `Image` pa ima i `GetWidth` i `GetHeight`.
I dalje ne vidim razlog da radite bitmapu neke druge veličine kad već u prkazu radite _Stretch_ - zadržite bitmapu (HBITMAP) originalne veličine, a promijenite veličinu prozora (kako god hoćete).

soperkov commented:
Kada izbacim GetObject odnosno one dvije metode za get width i height, a deklariram varijblu tipa bitmap unutar id_open casea i spremim njen sadrzaj u hbitmap, tada ne mogu pristupiti visini i sirini slike unutar on_paint-a. Koje bi tu onda bilo najbolje rješenje? Napraviti jednu bitmap varijablu u main window kao člana klase pa onda preko njega dobivati sve potrebno ili? Nisam sigurna, a dosta toga što sam probala ne radi kako spada pa evo ako može pomoć

ncaklovic commented:
Dodajte još dva membera za `width` i `height` i popunite ih dok imate `Bitmap`.

soperkov committed:
dodane nove varijable width i height, dodan unique_ptr

ncaklovic commented:
Da, to već vrlo liči na moje rješenje (sa `HBITMAP`). 
S malim razlikama:
- koristio sam lokalnu varijablu tipa `Bitmap` (umjesto `unique_ptr`) i konstruktor sa path-om
- nisam koristio `new_img_bmp` - ako pomaknete `DeleteObject` prije `GetHBITMAP` možete direktno `GetHBITMAP` u member 
- `SetStretchBltMode` sigurno treba na `hdc` (_destination_), ne na `memDC` (_source_)
- ne treba vam drugi `RectF` u `draw_string` ako `Offset` napravite između dva poziva `DrawString`

soperkov committed:
izmijenjeno ovako kako ste napisali
*SetStretchBltMode je bio razlog zašto su slike izgledale čudno, sada je puno bolje i jasnije.

ncaklovic commented:
Eto, kao što sam rekao - tri linije koda, par minuta posla :grin:
